[Lim Zhen Wy] iP

limzhenwy opened the pull request:
# Skog 
> "Your mind is for having ideas, not holding them." - David Allen [Source](https://dansilvestre.com/productivity-quotes/)

Skog frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need is
1. download it from [here](https://nus-cs2103-ay2324s1.github.io/website/schedule/week4/project.html).
1. double-click it.
1. add your tasks.
1. let it manage your tasks for you. :bear:

And it is **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (Coming soon)
- [ ] Reminders (Coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the main method:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

limzhenwy committed:
Altered chat bot name and changed the message

limzhenwy committed:
Added in the echo feature

limzhenwy committed:
List feature etc...

limzhenwy committed:
Mark and unmark features

limzhenwy committed:
Added all the 3 events and event logic

limzhenwy committed:
Exception handling

limzhenwy committed:
Delete function

limzhenwy committed:
Fixed some bugs and added method descriptors

limzhenwy committed:
Erronous commit and push in the last push

limzhenwy committed:
Separate individual classes into files

limzhenwy committed:
Add save function

limzhenwy committed:
Improve program logic

limzhenwy committed:
Add time feature

limzhenwy committed:
Make code more OOP

limzhenwy committed:
Make code more OOP

limzhenwy committed:
Fix bugs from OOP implementation

limzhenwy committed:
Merge branch 'branch-A-MoreOOP'

limzhenwy committed:
Put classes into duke package

limzhenwy committed:
Merge tag 'A-Packages'

limzhenwy committed:
Merge branch 'branch-A-Gradle' into add-gradle-support

limzhenwy committed:
Resolve unchecked errors

limzhenwy committed:
Merge branch 'add-gradle-support' into branch-A-Gradle

limzhenwy committed:
Merge branch 'branch-A-Gradle'

limzhenwy committed:
Add JUnit tests

limzhenwy committed:
Merge branch 'branch-A-JUnit'

limzhenwy committed:
Modify build.gradle to support JAR creation

limzhenwy committed:
Merge branch 'branch-A-Jar'

limzhenwy committed:
Add javadoc requirements

limzhenwy committed:
Merge branch 'branch-A-JavaDoc'

limzhenwy committed:
Add find feature

limzhenwy committed:
Merge branch 'branch-Level-9'

limzhenwy committed:
Edit code to follow convention

limzhenwy committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/duke/Event.java
#	src/main/java/duke/TaskList.java
#	src/main/java/duke/Todo.java

limzhenwy committed:
Fix bugs from merge conflict

jeffrey-jian commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,38 @@
+ * Special kind of task that has a description and a deadline
+ */
+class Deadline extends Task {
+    protected String type = "D";
```
Can consider using the `final` keyword for this variable!

```suggestion
    protected final String type = "D";
```

jeffrey-jian commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,108 @@
+                String[] arr = ui.sc.nextLine().split(DELIM);
+                String type = arr[0];
+
+                if (type.equals("bye")) {
+                    ui.end();
+                    storage.saveTextFile(tasks);
+                    storage.saveDataFile(tasks);
+                    System.out.println(EXIT);
+                    break;
+                } else if (type.equals("list")) {
+                    tasks.listOut();
+                } else if (type.equals("mark")) {
+                    tasks.mark(arr);
+                } else if (type.equals("unmark")) {
+                    tasks.unmark(arr);
+                } else if (type.equals("delete")) {
+                    tasks.delete(arr);
+                } else if (type.equals("find")) {
+                    tasks.find(arr);
+                } else {
+                    // check for task type first
+                    if (type.equals("todo")) {
+                        if (arr.length == 1) {
+                            throw new EmptyDescription();
+                        } else {
+                            String desc = tasks.getDescription(arr);
+                            tasks.addTodo(desc);
+                        }
+                    } else if (type.equals("deadline")) {
+                        String desc = tasks.getDescription(arr);
+                        String date = tasks.getDeadline(arr);
+                        tasks.addDeadline(desc, date);
+                    } else if (type.equals("event")) {
+                        String desc = tasks.getDescription(arr);
+                        String timeline = tasks.getEventTimeline(arr);
+                        tasks.addEvent(desc, timeline);
```
You may consider using switch-case statements for these if-else checks! Java automatically uses the String `.equals()` method when comparing Strings in switch-case statements.

jeffrey-jian commented on src/main/java/duke/EmptyDescription.java:
```diff
@@ -0,0 +1,16 @@
+    /**
+     * Constructor for the EmptyDescription class.
+     */
+    public EmptyDescription() {
```
Nice abstraction for `EmptyDescription` exceptions! Perhaps you can also include to use this `EmptyDescription` for other tasks (and not limited to toDos.

jeffrey-jian commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,257 @@
+     * String to show the user the current state of the Task ArrayList.
+     * @return String containing all information of the current Task ArrayList
+     */
+    public String totalTasks() {
```
> Use nouns for classes/variables and verbs for methods/functions.

```suggestion
    public String countTotalTasks() {
```

jeffrey-jian reviewed (commented):
LGTM!! Just a few nits to fix.

limzhenwy committed:
Update src/main/java/duke/TaskList.java

Co-authored-by: jeffrey-jian <74343245+jeffrey-jian@users.noreply.github.com>

limzhenwy committed:
Update src/main/java/duke/Deadline.java

Co-authored-by: jeffrey-jian <74343245+jeffrey-jian@users.noreply.github.com>

limzhenwy committed:
Fix test and bugs

limzhenwy committed:
Super duper ultra fun first step, great emphasis on super duper ultra FUN

limzhenwy committed:
Complete GUI up to part 4

limzhenwy committed:
Implement part 4 of GUI

limzhenwy committed:
Implement chatbot to handle commands

limzhenwy committed:
Merge branch 'branch-level-10'

Jweewee commented on src/test/java/duke/TaskListTest.java:
```diff
@@ -0,0 +1,29 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+public class TaskListTest {
+    @Test
+    public void nullArrayTest() {
```
Comprehensive Test case, looks good

Jweewee reviewed (commented):
LGTM

limzhenwy committed:
Add assertions to important parts of logic

limzhenwy committed:
Refactor code and improve code quality

Previously, the code was very messy and had inconsistent
levels of abstractions throughout.

Created different packages and moved classes into their
respective packages depending on their function.

This provides an easier way to read the code and check for
bugs and logic errors.

limzhenwy committed:
Add friendlier syntax

Typing out the entire command can be tedious and frustrating
when dealing with many tasks at hand.

Added multiple shorthands for users to be able to
easily specify the command type without having the need
to type out the entire command.

This allows users to be able to use the functions of the
chat bot much more easily and provides for a simpler
experience.

limzhenwy committed:
Update readme

limzhenwy committed:
Merge branch 'branch-A-UserGuide'

limzhenwy committed:
Update build.gradle for fat .jar file

limzhenwy committed:
Add Ui screenshot

limzhenwy committed:
Merge branch 'branch-A-Release'

limzhenwy committed:
Change comment

limzhenwy committed:
Merge pull request #2 from limzhenwy/branch-A-PR

Change comment

limzhenwy committed:
Add reply for task adding

limzhenwy committed:
Merge branch 'branch-reply'

limzhenwy committed:
Add UG

limzhenwy committed:
Merge branch 'branch-UG'

limzhenwy committed:
Improve formatting

limzhenwy committed:
Merge branch 'branch-UG'

limzhenwy committed:
Add runtime exception handling

limzhenwy committed:
Merge branch 'branch-ParseException'
